crmsh/migration.py

```diff
@@ -0,0 +1,655 @@
+    check_unsupported_resource_agents(handler, cib)
+
+
+def check_dependency_version(handler: CheckResultHandler):
+    handler.log_info('Checking dependency version...')
+    shell = sh.LocalShell()
+    out = shell.get_stdout_or_raise_error(None, 'corosync -v')
+    _check_version_range(
+        handler,
+        'Corosync', (3,),
+        re.compile(r"version\s+'(\d+(?:\.\d+)*)'"),
+        shell.get_stdout_or_raise_error(None, 'corosync -v'),
+    )
+
+
+def _check_version_range(
+        handler: CheckResultHandler, component_name: str,
+        minimum: tuple,
+        pattern,
+        text: str,
+):
+    match = pattern.search(text)
+    if not match:
+        handler.handle_problem(
+            True, handler.LEVEL_ERROR,
+            f'{component_name} version not supported',
+            [
+                'Unknown version:',
+                text,
+            ],
+        )
+    else:
+        version = tuple(int(x) for x in match.group(1).split('.'))
+        if not minimum <= version:
+            handler.handle_problem(
+                True,  handler.LEVEL_ERROR,
+                f'{component_name} version not supported', [
+                    'Supported version: {} <= {}'.format(
+                        '.'.join(str(x) for x in minimum),
+                        component_name,
+                    ),
+                    f'Actual version:    {component_name} == {match.group(1)}',
+                ],
+            )
+
+
+def check_unsupported_corosync_features(handler: CheckResultHandler):
+    handler.log_info("Checking used corosync features...")
+    conf_path = corosync.conf()
+    with open(conf_path, 'r', encoding='utf-8') as f:
+        config = corosync_config_format.DomParser(f).dom()
+    if config['totem'].get('rrp_mode', None) in {'active', 'passive'}:
+        handler.handle_problem(
+            False, handler.LEVEL_WARN,
+            'Corosync RRP will be deprecated in corosync 3.', [
+                'Run "crm health sles16 --fix" to migrate it to knet multilink.',
+            ],
+        )
+    _check_unsupported_corosync_transport(handler, config)
+
+
+def _check_unsupported_corosync_transport(handler: CheckResultHandler, dom):
+    transport = dom['totem'].get('transport', None)
+    if transport == 'knet':
+        return
+    if transport is None:
+        try:
+            dom['totem']['interface'][0]['bindnetaddr']
+        except KeyError:
+            # looks like a corosync 3 config
+            return
+    handler.handle_problem(
+        False, handler.LEVEL_WARN,
+        f'Corosync transport "{transport}" will be deprecated in corosync 3. Please use knet.', [],
+    )
+
+
+def migrate_corosync_conf():
+    conf_path = corosync.conf()
+    with open(conf_path, 'r', encoding='utf-8') as f:
+        config = corosync_config_format.DomParser(f).dom()
+    logger.info('Migrating corosync configuration...')
+    migrate_corosync_conf_impl(config)
+    shutil.copy(conf_path, conf_path + '.bak')
```

Comment:
crmsh leave corosync.conf.bak here, do not remove it after migration, or do not mention that this bak file has been created?